Fix #5811 - Clicking hint field blocks key input

david-allison-1 opened the pull request:
## Review Goals

* Confirm whether selecting the Text Input Field, rather than "Show Answer" on a note type with both `{{hint:}}` and `{{type:}}` is correct. 
* Normally, a `{{type:}}` field does not have this focus initially.
  * But, even with focus on the text input field, `ENTER` displays the new card.

* `_relinquishFocus()` is appropriate in cards.js, named correctly (and marked well as unstable). 
* Whether the extraction of some parsing logic is a good first step in taming the complexity of `filterUrl(String url)`, or if there's a better design.

## Purpose / Description

After clicking on a `{{hint:}}` field with the mouse, the WebView took focus and KeyEvents were not sent to the Activity.

The mouse cursor was stuck as a text entry bar when hovering over the WebView

## Fixes
Fixes #5811

## Approach

* We define a JavaScript function: `_relinquishFocus`. This navigates to a URL which we capture from the Activity.
* We call the above function onClick of the Hint field.
* We handle the signal in `AbstractFlashcardViewer` and remove focus from the WebVew
* We abstract out the concern of parsing the URL
  * Slightly reduces mental load when looking at the function. Readers now see the "What", not the "How".
  * Testability, and allows us to store known good values to call URL parsing.

## How Has This Been Tested?

Unit Tested

* Tested via Keyboard & Mouse with Template: 

```
{{Front}}
{{hint:Front}}
```

**Input:** Click the hint. Press "Enter"

**Output:** Show Answer is pressed.

* Tested via Keyboard & Mouse with Template:
* "Type answer into the card" is disabled

```
{{type:Front}}

{{hint:Back}}
```

**Input:** Click the hint. Press "A"

**Outcome:** "A" appears in the Android TextView.

## Learning

The WebView only appears to allows focus from touch events when "Type answer into the card" (via HTML input) is checked.

Even though this is the case, the WebView can be focused from other inputs, and this will need to be more closely looked at in future.

## Checklist
- [ X ] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [ X ] You have a descriptive commit message with a short title (first line, max 50 chars).
- [ X ] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [ X ] You have commented your code, particularly in hard-to-understand areas
- [ X ] You have performed a self-review of your own code

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5179246&bid=16997852)

timrae commented on AnkiDroid/src/main/assets/scripts/card.js:
```diff
@@ -58,6 +58,17 @@ var resizeImages = function() {
+function _relinquishFocus() {
+    // Clicking on a hint set the Android mouse cursor to a text entry bar, even after navigating
+    // away. This fixes the issue.
+    document.body.style.cursor = 'default';
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Strings must use doublequote.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5179246)

mikehardy replied:
I don't believe we have a solid javascript code standard, I haven't checked to see if this is following convention but if it is, it can be ignored

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -2925,4 +2944,24 @@ protected void dismiss(Collection.DismissType type) {
+
+    /** Signals from a WebView represent actions with no parameters */
+    @VisibleForTesting
+    static class WebViewSignalParser {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The utility class name 'WebViewSignalParser' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper)'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5179246)

timrae commented on AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java:
```diff
@@ -8,6 +8,7 @@
 
 import androidx.test.ext.junit.runners.AndroidJUnit4;
 
+import static com.ichi2.anki.AbstractFlashcardViewer.WebViewSignalParser.*;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'com.ichi2.anki.AbstractFlashcardViewer.WebViewSignalParser'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5179246)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5179246&bid=16997852)

timrae commented on AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java:
```diff
@@ -8,6 +8,7 @@
 
 import androidx.test.ext.junit.runners.AndroidJUnit4;
 
+import static com.ichi2.anki.AbstractFlashcardViewer.WebViewSignalParserUtils.*;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'com.ichi2.anki.AbstractFlashcardViewer.WebViewSignalParserUtils'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5179246)

david-allison-1 replied:
This is used.

Could be an issue with the auxclasspath: https://github.com/pmd/pmd/issues/1870

mikehardy commented:
I like this - I consider my comments nits for the most part. I'm curious for Tim's thoughts but in general I'd merge pending responses, thanks again!

david-allison-1 commented:
@mikehardy I'm missing your comments. I can see:  `@ankidroid ankidroid deleted a comment from mikehardy 11 hours ago`, which may be related to the fact that I amended/force pushed a few spelling issues.

mikehardy commented:
@david-allison-1 indeed - if my comment was codacy-related (and I am 99% sure this one was, about quote style in javascript) and you force-push something related it nukes them. No worries

mikehardy commented:
As I was sweeping through long lost browser tabs I just realized I had not committed my review comments yet - did that just now

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5179246&bid=16997852)

timrae commented on AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java:
```diff
@@ -8,6 +8,7 @@
 
 import androidx.test.ext.junit.runners.AndroidJUnit4;
 
+import static com.ichi2.anki.AbstractFlashcardViewer.WebViewSignalParserUtils.*;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'com.ichi2.anki.AbstractFlashcardViewer.WebViewSignalParserUtils'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5179246)

mikehardy replied:
@david-allison-1 no worries on things that are clear like this. We don't consider codacy failures a hard stop for a variety of reasons, one of which is that you can fix 9 items in a file with 10, and it'll flag you for the 10th even though it's unrelated to the PR. I always take Codacy into consideration but don't block on it

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -1651,6 +1660,16 @@ protected void hideEaseButtons() {
         mEase3Layout.setVisibility(View.GONE);
         mEase4Layout.setVisibility(View.GONE);
         mFlipCardLayout.setVisibility(View.VISIBLE);
+        focusAnswerCompletionField();
```
Related and to consider: #4699

david-allison-1 commented:
My comment seems to have been deleted by Codacy: There may be an issue with PMD and  `auxclasspath` on the unit tests jar: `https://github.com/pmd/pmd/issues/1870`. The import is required.

david-allison-1 commented:
@mikehardy I'd like to revert the previous commit as there's a logic error. 

While reviewing in this repo, is it preferable to perform a revert operation, or rebase to `HEAD~1` and force push?

mikehardy commented:
We are a "clean commit history" repo not a "don't alter history" repo, we consider the final work the thing to commit, not the trail of work over time. So it's also rebase HEAD~N + force push

mikehardy commented:
@david-allison-1 I knew #5817 would blow this one up with a conflict but I also knew you would probably only take 5mins to fix it so I merged that first and put the conflict on you, sorry 🙏

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5179246&bid=16997852)

timrae commented on AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java:
```diff
@@ -8,6 +8,7 @@
 
 import androidx.test.ext.junit.runners.AndroidJUnit4;
 
+import static com.ichi2.anki.AbstractFlashcardViewer.WebViewSignalParserUtils.*;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'com.ichi2.anki.AbstractFlashcardViewer.WebViewSignalParserUtils'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5179246)

timrae replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'com.ichi2.anki.AbstractFlashcardViewer.WebViewSignalParserUtils'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5179246)

david-allison-1 commented:
Merged, rebased and force pushed.

Note that I've moved the logic added in the above commit into `getSignalFromUrl`, and added additional tests. I think this simplifies `getSignalFromUrl`.

Main goals: 
* Ensure there is no logical change from the answer & ordinal browser signal processing
* Ensure you're happy with naming

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -3092,4 +3113,43 @@ protected void dismiss(Collection.DismissType type) {
+            }
+
+            if (url.startsWith("signal:answer_ease")) {
+                return SIGNAL_HANDLED;
```
I don't understand how a signal that starts with answer_ease is handled, when the 4 we expect are above (and are handled in a switch when they are seen by the caller) but handled is actually not handled? who is handling it? Might be naming semantics - might be better as unhandled with a little more logging about an answer_ease value that we're not expecting?

david-allison-1 replied:
Yep, naming. Will consider renaming

Currently 
Unhandled: We don't know about this, it's not a signal so try intent processing.
Handled: We know about this, but have no associated action to take.

Maybe `SIGNAL_HANDLED` => `SIGNAL_NOOP`?

david-allison-1 replied:
I guess the aim is: we want to stop execution and return true from a NOOP, and continue processing from anything unhandled.

mikehardy replied:
SIGNAL_NOOP sounds good, with an associated Timber.w because it feels like it is unexpected input/state (thus more than info), but we can certainly continue processing (thus not error) ?

mikehardy commented:
Travis feels overloaded at the moment (maybe we're all working on open source stuff while quarantined?) so it's failing spuriously for this and other builds, I'm on it though - kicking builds off again when they need it

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5179246&bid=16997852)
